components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -1019,4 +1021,165 @@ GURL AssetDiscoveryManager::GetSimpleHashNftsByWalletUrl(
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    return;
+  }
+
+  auto user_assets = BraveWalletService::GetUserAssets(prefs_);
+  std::vector<std::string> user_assets_chain_ids;
+  std::transform(
+      user_assets.begin(), user_assets.end(),
+      std::back_inserter(user_assets_chain_ids),
+      [](const mojom::BlockchainTokenPtr& token) { return token->chain_id; });
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          user_assets_chain_ids);
+
+  base::flat_map<std::string, std::vector<std::string>>
+      chain_id_to_contract_addresses;
+
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    // Skip all except only next allowed networks,
+    // until requests across all blocks will not
+    // be allowed for the RPC method "eth_getLogs"
+    if (chain_id != mojom::kMainnetChainId &&
+        chain_id != mojom::kPolygonMainnetChainId &&
+        chain_id != mojom::kAvalancheMainnetChainId &&
+        chain_id != mojom::kCeloMainnetChainId &&
+        chain_id != mojom::kArbitrumMainnetChainId &&
+        chain_id != mojom::kOptimismMainnetChainId) {
+      continue;
+    }
+
+    for (const auto& token : token_list) {
+      if (token->coin != mojom::CoinType::ETH) {
+        continue;
+      }
+      chain_id_to_contract_addresses[chain_id].push_back(
+          token->contract_address);
+    }
+  }
+
+  if (chain_id_to_contract_addresses.empty()) {
+    return;
+  }
+  // Use a barrier callback to wait for all EthGetLogs calls to
+  const auto barrier_callback =
+      base::BarrierCallback<std::vector<mojom::AllowanceInfoPtr>>(
+          account_addresses.size() * chain_id_to_contract_addresses.size(),
+          base::BindOnce(&AssetDiscoveryManager::MergeEthAllowances,
+                         weak_ptr_factory_.GetWeakPtr()));
+
+  const auto approvalTopicHash =
+      KeccakHash("Approval(address,address,uint256)");
+  for (const auto& account_address : account_addresses) {
+    for (const auto& [chain_id, contract_addresses] :
+         chain_id_to_contract_addresses) {
+      base::Value::List contracts;
+      for (const auto& contract_address : contract_addresses) {
+        contracts.Append(contract_address);
+      }
+
+      auto internal_callback =
+          base::BindOnce(&AssetDiscoveryManager::OnGetAllowances,
+                         weak_ptr_factory_.GetWeakPtr(), barrier_callback);
+
+      base::Value::Dict filter_options;
+      base::Value::List topics;
+      topics.Append(approvalTopicHash);
+      std::string hexZeroPadAccAddr;
+      if (!PadHexEncodedParameter(account_address, &hexZeroPadAccAddr)) {
+        continue;
+      }
+      topics.Append(std::move(hexZeroPadAccAddr));
+
+      filter_options.Set("address", std::move(contracts));
+      filter_options.Set("topics", std::move(topics));
+      filter_options.Set("fromBlock", "earliest");
```

Comment:
Maybe it's better that we track the fromBlock per chain and use it in future calls and not always query from the earliest blocks?
We might need to store allowance infos we discovered before with its block_number tho.

Reply:
Implemented it to cache block number in the preferences